src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+            if (this.ShouldLimitTimeouts())
+            {
+                long pendingTimeoutsCount = Interlocked.Increment(ref this.PendingTimeouts);
+                if (pendingTimeoutsCount > this.maxPendingTimeouts)
+                {
+                    Interlocked.Decrement(ref this.PendingTimeouts);
+                    throw new RejectedExecutionException($"Number of pending timeouts ({pendingTimeoutsCount}) is greater than or equal to maximum allowed pending timeouts ({this.maxPendingTimeouts})");
+                }
+            }
+
+            this.Start();
+
+            // Add the timeout to the timeout queue which will be processed on the next tick.
+            // During processing all the queued HashedWheelTimeouts will be added to the correct HashedWheelBucket.
+            TimeSpan deadline = CeilTimeSpanToMilliseconds((PreciseTimeSpan.Deadline(delay) - this.StartTime).ToTimeSpan());
+            var timeout = new HashedWheelTimeout(this, task, deadline);
+            this.timeouts.TryEnqueue(timeout);
+            return timeout;
+        }
+
+        bool ShouldLimitTimeouts() => this.maxPendingTimeouts > 0;
+
+        static void ReportTooManyInstances() =>
+            Logger.Error($"You are creating too many {nameof(HashedWheelTimer)} instances. {nameof(HashedWheelTimer)} is a shared resource that must be reused across the process,so that only a few instances are created.");
+
+        static TimeSpan CeilTimeSpanToMilliseconds(TimeSpan time)
+        {
+            long remainder = time.Ticks % TimeSpan.TicksPerMillisecond;
+            return remainder == 0 ? time : new TimeSpan(time.Ticks - remainder + TimeSpan.TicksPerMillisecond);
+        }
+
+        sealed class Worker : IRunnable
+        {
+            readonly HashedWheelTimer owner;
+            readonly ISet<ITimeout> unprocessedTimeouts = new HashSet<ITimeout>();
+
+            long tick;
+
+            public Worker(HashedWheelTimer owner)
+            {
+                this.owner = owner;
+            }
+
+            public void Run()
+            {
+                try
+                {
+                    // Initialize the startTime.
+                    this.owner.StartTime = PreciseTimeSpan.FromStart;
+                    if (this.owner.StartTime == PreciseTimeSpan.Zero)
+                    {
+                        // We use 0 as an indicator for the uninitialized value here, so make sure it's not 0 when initialized.
+                        this.owner.StartTime = PreciseTimeSpan.FromTicks(1);
+                    }
+
+                    // Notify the other threads waiting for the initialization at start().
+                    this.owner.startTimeInitialized.Signal();
+
+                    do
+                    {
+                        TimeSpan deadline = this.WaitForNextTick();
+                        if (deadline > TimeSpan.Zero)
+                        {
+                            int idx = (int)(this.tick & this.owner.mask);
```

Comment:
yes

---
In reply to: [114475721](https://github.com/Azure/DotNetty/pull/242#discussion_r114475721) [](ancestors = 114475721)